mathematica/integrator_plots.cpp

```diff
@@ -169,15 +172,148 @@ std::vector<SimpleHarmonicMotionPlottedIntegrator> ReferenceMethods() {
+      v_error_data.emplace_back(
+          PlottableDataset(evaluations_[i], v_errors_[i]));
+      e_error_data.emplace_back(
+          PlottableDataset(evaluations_[i], e_errors_[i]));
+      names.emplace_back(Escape(methods_[i].name));
+    }
+    std::string result;
+    result += Assign("qErrorData", q_error_data);
+    result += Assign("vErrorData", v_error_data);
+    result += Assign("eErrorData", e_error_data);
+    result += Assign("names", names);
+    return result;
+  }
+
+ private:
+  Status Integrate(int method_index, int time_step_index) {
+    auto const& method = methods_[method_index];
+    Problem problem;
+    int number_of_evaluations = 0;
+    problem.equation.compute_acceleration = std::bind(
+        compute_accelerations_, _1, _2, _3, &number_of_evaluations);
+    problem.initial_state = &initial_state_;
+    auto const t0 = problem.initial_state->time.value;
+    Length max_q_error;
+    Speed max_v_error;
+    Energy max_e_error;
+    auto append_state = [this, &max_q_error, &max_v_error, &max_e_error](
+      ODE::SystemState const& state) {
+      auto const errors = compute_errors_(state);
+      max_q_error = std::max(max_q_error, errors.q_error);
+      max_v_error = std::max(max_v_error, errors.v_error);
+      max_e_error = std::max(max_e_error, errors.e_error);
+    };
+    Time const Δt = method.evaluations * starting_step_size_per_evaluation_ /
+                    std::pow(step_reduction_, time_step_index);
+    auto instance = method.integrator.NewInstance(problem, append_state, Δt);
+
+    instance->Solve(tmax_);
+    // Log both the actual number of evaluations and a theoretical number
+    // that ignores any startup costs; that theoretical number is the one
+    // used for plotting.
+    int const amortized_evaluations =
+        method.evaluations * static_cast<int>(std::floor((tmax_ - t0) / Δt));
+    LOG_EVERY_N(INFO, 50) << "[" << method_index << "," << time_step_index
+                          << "] " << problem_name_ << ": "
+                          << number_of_evaluations << " actual evaluations ("
+                          << amortized_evaluations << " amortized) with "
+                          << method.name;
+    // We plot the maximum error, i.e., the L∞ norm of the error.
+    // Blanes and Moan (2002), or Blanes, Casas and Ros (2001) tend to use
+    // the average error (the normalized L¹ norm) instead.
+    q_errors_[method_index][time_step_index] = max_q_error;
+    v_errors_[method_index][time_step_index] = max_v_error;
+    e_errors_[method_index][time_step_index] = max_e_error;
+    evaluations_[method_index][time_step_index] = amortized_evaluations;
+
+    return Status::OK;
+  }
+
+  std::vector<SimpleHarmonicMotionPlottedIntegrator> const methods_;
+  std::function<void(Instant const& t,
+                     std::vector<Length> const& q,
+                     std::vector<Acceleration>& result,
+                     int* evaluations)>
```

Comment:
Then it should really be `std::optional<int>& evaluations`.  Please put a TODO in `testing_utilities`.